modules/schema/src/main/java/org/apache/ignite/internal/schema/TableRow.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.schema;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+/** Heap byte buffer-based row. */
+public class TableRow {
+    public static final ByteOrder ORDER = ByteOrder.LITTLE_ENDIAN;
+
+    /** Size of schema version length field. */
+    private static final int SCHEMA_VERSION_FLD_LEN = Short.BYTES;
+
+    /** Row schema version field offset. */
+    private static final int SCHEMA_VERSION_OFFSET = 0;
+
+    /** Row binary tuple field offset. */
+    private static final int TUPLE_OFFSET = SCHEMA_VERSION_OFFSET + SCHEMA_VERSION_FLD_LEN;
+
+    /** Row buffer. */
+    private final ByteBuffer buf;
+
+    /**
+     * Constructor.
+     *
+     * @param data Array representation of the row.
+     */
+    public TableRow(byte[] data) {
+        this(ByteBuffer.wrap(data).order(ORDER));
+    }
+
+    /**
+     * Constructor.
+     *
+     * @param buf Buffer representing the row.
+     */
+    public TableRow(ByteBuffer buf) {
+        assert buf.order() == ORDER;
+        assert buf.position() == 0;
+
+        this.buf = buf;
+    }
+
+    /**
+     * Factory method which creates an instance using a schema version and an existing {@link BinaryTuple}.
+     *
+     * @param schemaVersion Schema version.
+     * @param binaryTuple Binary tuple.
+     * @return Created TableRow instance.
+     */
+    public static TableRow createFromTuple(int schemaVersion, BinaryTuple binaryTuple) {
```

Comment:
What is the reason of encoding schema version directly into the binary payload? This leads to too much memory manipulation. Storages can do that more efficiently.
I suggest storing version it as an integer field, and only assemble a byte[] when asked.

Reply:
> What is the reason of encoding schema version directly into the binary payload? 

it is a direct replacement for ByteBufferRow to facilitate migration

>  I suggest storing version as an integer field

if the storage doesn't mind to work not with abstract bytes, but with a specific row object, let's do this way

Reply:
I think we should make these migrations as simple as possible and avoid making changes not relevant to the binary format.

Reply:
Ok, but we should fix this in the future. There must be a TODO. Current approach is not efficient